spring-cloud-gcp-autoconfigure/pom.xml

```diff
@@ -263,7 +263,6 @@
         <dependency>
             <groupId>com.google.cloud</groupId>
             <artifactId>spring-cloud-gcp-secretmanager</artifactId>
-            <optional>true</optional>
```

Comment:
Can't do that because it will turn on the autoconfiguration even when the secrets manager starter is not imported.

Reply:
By adding `spring.config.import=sm://` implicitly to property files, Spring Boot will try to register secret-manager related beans and get the following error:
>java.lang.NoClassDefFoundError: com/google/cloud/secretmanager/v1/SecretManagerServiceClient

	at com.google.cloud.spring.autoconfigure.secretmanager.SecretManagerConfigDataLocationResolver.registerSecretManagerBeans(SecretManagerConfigDataLocationResolver.java:71)
	at com.google.cloud.spring.autoconfigure.secretmanager.SecretManagerConfigDataLocationResolver.resolve(SecretManagerConfigDataLocationResolver.java:60)

This is because `spring-cloud-gcp-secretmanager` dependency doesn't in the classpath.

Reply:
Maybe `isResolvable` method in `SecretManagerConfigDataLocationResolver` should return `false` if `spring-cloud-gcp-secretmanager` is not in the classpath.

Reply:
I don't know if we should try to add `spring.config.import=sm://` implicitly. Can we just register a `SecretManagerTemplate` bean in the application context? Can we use the regular auto-configuration, or is there some issue because we're dealing bootstrap context?

Reply:
When we use Spring Config API (thus deprecate bootstrap phase), all the Secret Manager related beans creation steps move to `SecretManagerConfigDataLocationResolver.resolve()` which only gets call if `spring.config.import=sm://` presents.

If we register a `SecretManagerTemplate` bean using auto-configuration, it's essentially brought back `GcpSecretManagerBootstrapConfiguration`.

Reply:
Kind of, but it doesn't need to be bootstrap. I was thinking something like this: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/1567.

The beans created with `SecretManagerConfigDataLocationResolver` would still take precedence over `GcpSecretManagerAutoConfiguration`, but if the resolver is not used, the `SecretManagerTemplate` bean would be created by the autoconfig. WDYT?

Reply:
Sure, this might be possible.

Would you like me to close this PR and start from yours?

Reply:
> By adding spring.config.import=sm:// implicitly to property files, Spring Boot will try to resolve Config Data Locations in ConfigDataEnvironmentPostProcessor, then SecretManagerConfigDataLocationResolver trys to register beans which eventually gets the following error:

Can you explain a bit more why there is a difference between adding the property implicitly vs. explicitly? Wouldn't `SecretManagerConfigDataLocationResolver` be activated either way?

Reply:
In current implementation, `SecretManagerConfigDataLocationResolver.resolve()` will not be called if no `spring.config.import` exists.

By adding `spring.config.import` explicitly, we expect customers to add secret manager starter dependency since they want to use secret manager. In this case, `SecretManagerConfigDataLocationResolver.resolve()` is called.

However, if we add `spring.config.import` implicitly, i.e., adding it in `GcpSecretManagerEnvironmentPostProcessor`, `SecretManagerConfigDataLocationResolver.resolve()` will always be called so secret manager starter dependency is required in classpath.

Reply:
Got it, thanks!